### PR TITLE
[Merged by Bors] - Implement init_resource for `Commands` and `World`

### LINKED ISSUE #1397: Add non-send resources via Commands
**What problem does this solve or what need does it fill?**

I have a non-send resource ([`RustPython`](https://github.com/RustPython/RustPython) interpreter). It has a sufficiently complicated construction process such that I didn't want to run the constructor in `Plugin::build`, but rather have my plugin launch a startup system that would insert the non-send resource. 

Right now, the only way to insert a non-send (thread-local) resource is either via `AppBuilder`, or via a thread-local system that directly manipulates `Resources`.

**What solution would you like?**

Using the `Commands` struct provides a more consistent way of managing resources outside of `AppBuilder`. As proposed by RawMeatball [here](https://discord.com/channels/691052431525675048/742569353878437978/806918993763237938), the API could look like:

```rust
fn forced_non_send_system(
  commands: &mut Commands,
  marker: NonSend<()>,
) {
  let custom_thing = Stuff::that_is_non_send();
  commands.insert_non_send_resource(&marker, custom_thing);
}
```

**What alternative(s) have you considered?**

Could just leave it the way it is. This isn't a necessity, but a nice bit of API consistency.

### LINKED ISSUE #3078: Add Commands::init_resource
## What problem does this solve or what need does it fill?

FromWorld resources cannot be added via commands without using a custom command.

## What solution would you like?

Implement `Commands::init_resource`, which is a trivial wrapper over `World::init_resource`.

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- Fixes #3078
- Fixes #1397

## Solution

- Implement Commands::init_resource.
- Also implement for World, for consistency and to simplify internal structure.
- While we're here, clean up some of the docs for Command and World resource modification.

### CODE REVIEW

**Comment by DJMcNab** [comment]
I don't like that `App::init_resource` doesn't overwrite, but `World::init_resource` does overwrite. Is there a reason you've chosen different behaviour for these methods?

As of https://github.com/bevyengine/bevy/pull/1349 (which you were in favour of 👀)  `App::init_resource` doesn't overwrite:
https://github.com/bevyengine/bevy/blob/225d6a138f8f5be92ef42ce146a76d4c43f60f72/crates/bevy_app/src/app.rs#L667-L679
(which is current `main`)

**Comment by alice-i-cecile** [comment]
> I don't like that App::init_resource doesn't overwrite, but World::init_resource does overwrite. Is there a reason you've chosen different behaviour for these methods?

I had forgotten that #1349 landed, and wrote this code from scratch ;) Let me fix that.

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -591,64 +591,104 @@ impl World {
         }
     }
 
+    /// Inserts a new resource with default values.
+    ///
+    /// If the resource already exists, nothing happens.
+    ///
+    /// Uses the [`FromWorld`] trait to determine values,
+    /// which has a blanket impl for all `T: Default`.
```
Comment (NiklasEi) [inline crates/bevy_ecs/src/world/mod.rs:599]:
```suggestion
    /// which has a blanket impl for all `R: Default`.
```
For me, this would fit better since we use R as the generic. If you agree, please do the same change for `init_non_send_resource`.

File: crates/bevy_ecs/src/world/mod.rs [hunk 2]
```diff
@@ -591,64 +591,104 @@ impl World {
...
+        if !self.contains_resource::<R>() {
+            let resource = R::from_world(self);
+            self.insert_resource(resource);
+        }
+    }
+
     /// Inserts a new resource with the given `value`.
+    ///
     /// Resources are "unique" data of a given type.
+    /// Insert a resource of a type that already exists
```
Comment (NiklasEi) [inline crates/bevy_ecs/src/world/mod.rs:614]: This reads a bit weird. Maybe "If a resource of type `R` already exists, its value will be overwritten."?

File: crates/bevy_ecs/src/world/mod.rs [hunk 3]
```diff
@@ -591,64 +591,104 @@ impl World {
...
+    /// will overwrite any existing data.
     #[inline]
     pub fn insert_resource<T: Resource>(&mut self, value: T) {
         let component_id = self.components.init_resource::<T>();
         // SAFE: component_id just initialized and corresponds to resource of type T
         unsafe { self.insert_resource_with_id(component_id, value) };
     }
 
+    /// Inserts a new resource with default values.
```
Comment (NiklasEi) [inline crates/bevy_ecs/src/world/mod.rs:623]:
```suggestion
    /// Inserts a new non-send resource with default values.
```

**Review by NiklasEi — APPROVED** [review]
Nice, I like it.

Added migration guide label due to the renamed world method.

File: crates/bevy_ecs/src/world/mod.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/world/mod.rs:599]: Ah, I missed this one!

File: crates/bevy_app/src/app.rs [hunk 4]
```diff
@@ -631,19 +628,16 @@ impl App {
...
-        T: 'static,
-    {
-        self.world.insert_non_send(resource);
+    pub fn insert_non_send_resource<R: 'static>(&mut self, resource: R) -> &mut Self {
+        self.world.insert_non_send_resource(resource);
         self
     }
 
-    /// Initialize a resource in the current [`App`], if it does not exist yet
+    /// Initialize a resource with default values by adding it to the [`World`]
```
Comment (mockersf) [inline crates/bevy_app/src/app.rs:636]: the values may come not from `Default` but from `FromWorld`, could you rephrase with another word than "default"?

File: crates/bevy_app/src/app.rs [hunk 5]
```diff
@@ -631,19 +628,18 @@ impl App {
...
-        T: 'static,
-    {
-        self.world.insert_non_send(resource);
+    pub fn insert_non_send_resource<R: 'static>(&mut self, resource: R) -> &mut Self {
+        self.world.insert_non_send_resource(resource);
         self
     }
 
-    /// Initialize a resource in the current [`App`], if it does not exist yet
+    /// Initialize a resource with standard starting values by adding it to the [`World`]
```
Comment (jamesbeilby) [inline crates/bevy_app/src/app.rs:636]: Minor suggestion to avoid the word "initialize" as it's often ambiguous whether it's separate to creation; perhaps instead:
> Create a resource with its standard starting values and insert it into the [`World`]
Or:
> Ensure a resource of the given type exists in the [`World`]. If it does not, insert a new instance of this resource with its standard starting values.

**Review by MinerSebas — APPROVED** [review]

**Review by cart — APPROVED** [review]
Looks good to me! Just resolve conflicts and we're good to go.

**Comment by alice-i-cecile** [comment]
@cart We're rebased now :)

File: crates/bevy_ecs/src/world/mod.rs [hunk 6]
```diff
@@ -592,64 +592,106 @@ impl World {
...
+    /// Inserts a new resource with standard starting values.
+    ///
+    /// If the resource already exists, nothing happens.
+    ///
+    /// The value given by the [`FromWorld::from_world`] method will be used.
+    /// Note that any resource with the `Default` trait automatically implements `FromWorld`,
+    /// and those default values will be here instead.
+    #[inline]
+    pub fn init_resource<R: Resource + FromWorld>(&mut self) {
+        // PERF: We could avoid double hashing here, since the `from_resources` call is guaranteed
```
Comment (bjorn3) [inline crates/bevy_ecs/src/world/mod.rs:604]:
```suggestion
        // PERF: We could avoid double hashing here, since the `from_world` call is guaranteed
```

File: crates/bevy_ecs/src/world/mod.rs [hunk 7]
```diff
@@ -592,64 +592,106 @@ impl World {
...
+    /// Inserts a new non-send resource with standard starting values.
+    ///
+    /// If the resource already exists, nothing happens.
+    ///
+    /// The value given by the [`FromWorld::from_world`] method will be used.
+    /// Note that any resource with the `Default` trait automatically implements `FromWorld`,
+    /// and those default values will be here instead.
+    #[inline]
+    pub fn init_non_send_resource<R: 'static + FromWorld>(&mut self) {
+        // PERF: We could avoid double hashing here, since the `from_resources` call is guaranteed
```
Comment (bjorn3) [inline crates/bevy_ecs/src/world/mod.rs:634]:
```suggestion
        // PERF: We could avoid double hashing here, since the `from_world` call is guaranteed
```

File: crates/bevy_ecs/src/world/mod.rs (hunk 6 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/world/mod.rs]: The corrections are certainly less wrong than the existing comments though :p
